pageserver/src/layered_repository.rs

```diff
@@ -1232,6 +1289,7 @@ impl LayeredTimeline {
             prev_record_lsn: ondisk_prev_record_lsn,
             ancestor_timeline: ancestor_timelineid,
             ancestor_lsn: self.ancestor_lsn,
+            current_logical_size: self.get_current_logical_size(),
```

Comment:
This can go out of sync on page server restarts. On restart, we restart decoding the WAL at `on_disk_consistent_lsn`. If there are any records that extend a segment between `on_disk_consistent_lsn`, they might get double-counted in `current_logical_size`. We would have updated `current_logical_size` before the restart already, and will do so again after restart.

In other words, `put_wal_record` and `put_page_image` should be idempotent.

Reply:
We write metadata with `current_logical_size` on checkpoints, the same time `on_disk_consitent_lsn` is updated. On restart `current_logical_size` (with all changes made by records after `on_disk_consistent_lsn`) is lost. And it is restored from metadata in consistent state with `on_disk_consitent_lsn`, and all changes will be restored when we reprocess WAL starting from `on_disk_consistent_lsn` and `corresponding current_logical_size`. Am I missing something? 

I think this design of size tracking can be improved, the problem I see is that `current_logical_size` is updated independently of advancing `last_record_lsn`, which I think can lead to errors and unclear in general. So the solution might be to split a pending value and apply it with corresponding advance calls. What do you think about that?

Reply:
> We write metadata with `current_logical_size` on checkpoints, the same time `on_disk_consitent_lsn` is updated. On restart `current_logical_size` (with all changes made by records after `on_disk_consistent_lsn`) is lost. And it is restored from metadata in consistent state with `on_disk_consitent_lsn`, and all changes will be restored when we reprocess WAL starting from `on_disk_consistent_lsn` and `corresponding current_logical_size`. Am I missing something?

You're saving the `current_logical_size` as it was at the *end* of the function call in the metadata, while `on_disk_consistent_lsn` usually points to some older LSN. Imagine this sequence:

0. System is running normally with current_logical_size = 1000, on_disk_consistent_lsn=100, last_record_lsn=200
1. Checkpoint begins and reads last_record_lsn as 200.
2. Relation is extended by 1 block at LSN 300, current_logical_size is now 1001.
3. Checkpoint runs and updates metadata with current_logical_size=1001, and on_disk_consistent_lsn with 200.
4. The server is restarted. current_logical_size=1001 is read from the metadata.
5. The WAL record at LSN 300 is replayed again. current_logical_size is incremented again, to 1002.

> I think this design of size tracking can be improved, the problem I see is that `current_logical_size` is updated independently of advancing `last_record_lsn`, which I think can lead to errors and unclear in general. So the solution might be to split a pending value and apply it with corresponding advance calls. What do you think about that?

Hmm, not sure how that would work.

I'd love to have some mechanism to recalculate the total size "from scratch", like get_current_logical_size_non_incremental() does. Currently that's only used for testing, but it could be used to resynchronize if the value becomes wrong for any reason. It seems inevitable that value goes out of sync eventually if we rely purely on "dead reckoning", whether it's due to some cumulative error or a bug. If we had that, we could use it to initialize the value at system startup, or perhaps knowingly start up with an out-of-date value but launch a background job to recalculate it. How accurate do we need the value to be?

Reply:
Hmm, I think I've got it, looks like kind of a race. Could this be fixed by remembering current logical size at the same time checkpoint reads last_record_lsn?
> Checkpoint begins and reads last_record_lsn as 200.

Agree that there needs to be backup mechanism.
> launch a background job to recalculate it
I think this is the best option because this way it won't affect startup time which is important for us.

> How accurate do we need the value to be?
It is supposed to be used for size quota tracking, so aggregate value for all timelines (without counting ancestors multiple times) is used to check against remaining space according to user's reserved capacity. In earlier discussions I remember we talked about precision and decided that the value could be approximate, but I think there should be a note on bulk changes, so if user truncates a relation to free some space we should reflect this rather quickly

> Hmm, not sure how that would work.
This could be done via current_logical_size which is advanced only in advance_last_record_lsn, and a pending_current_logical_size which is changed in for example put_page_version. When advance_last_record_lsn is called, pending part is combined with current, so current value reflects situation at last_record_lsn. There should be some locking for that co be consistent... Maybe this machinery is not needed at all and this is just redundant complexity

Reply:
@hlinnaka I've added remembering logical size before checkpoint and additional size equality assertions in `LayeredRepo::get_timeline_locked`, please take a look

Reply:
> @hlinnaka I've added remembering logical size before checkpoint and additional size equality assertions in `LayeredRepo::get_timeline_locked`, please take a look

Hmm, I don't think that's still quite right. `current_logical_size` holds to the size at `last_record_lsn`, but the `disk_consistent_lsn` value that we write to disk is usually older than `last_record_lsn`.

Working through another slightly different example:

1. System is running normally with current_logical_size = 1000
2. Relation is extended by 1 block at LSN 300, size of the database is now 1001.
3. Relation is extended again by 1 block at LSN 400, size of the database is now 1002 
4. Checkpoint begins and reads last_record_lsn as 400, and current_logical_size=1002.
5. Checkpoint runs, freezes some layers but not everything, and stops with on_disk_consistent=300. It updates metadata with with current_logical_size=1002 and on_disk_consistent_lsn=300.
6. The server is restarted. WAL decoding starts at LSN 300, with current_logical_size=1002.
7. The WAL record at LSN 400 is replayed again. current_logical_size is incremented to 1003.

Reply:
Hmm, I think I got it. What do you think about not storing incremental size on disk and load it on timeline start, and then begin using incremental updates? This can slightly slow down timeline loading, not sure how critical it is (probably not so critical for small sizes). Thoughts?

Reply:
@hlinnaka up

Reply:
Yeah, I think that's OK. And we can improve on that later if we need to.

Reply:
That's been already implemented in the latest commit, so please take a look :)